cups-pdf(-to-pdf): init

Yarny0 opened the pull request:
###### Description of changes

Implements the package request in https://github.com/NixOS/nixpkgs/issues/173745

cups-pdf is a cups printer backend that turns print jobs into pdf files.

* This does not package cups-pdf from https://www.cups-pdf.de/, but a fork cups-pdf-to-pdf from https://github.com/alexivkin/CUPS-PDF-to-PDF .  It claims to try harder to preserve searchable text in pdf files where the original cups-pdf rasterizes pdf files producing large pdf files with embedded images.

  https://github.com/alexivkin/CUPS-PDF-to-PDF/blob/c14428c2ca8e95371daad7db6d11c84046b1a2d4/README.md
  https://bugs.launchpad.net/ubuntu/+source/cups-pdf/+bug/366949

* A NixOS module is included that, if enabled, installs a cups-pdf printer backend and a cups queue for that printer so that one can print to pdf without further ado.  The module is quite flexible: It permits to install several cups-pdf backend "instances" (so one can use different configurations), and it can be told *not* to install a cups queue, so the user can install one or even several queues (again with varying configurations) if needed.

  Note that cups calls its backend processes with an ordinary user account, but cups-pdf wants to be called with root privileges because it assigns the print job's owner as ownership to the produced files.  Since we don't want to change the cups configuration so that it calls every program with root privileges, this NixOS module installs a suid root wrapper.  The wrapper is configured such that only cups (only the `lp` group) can call it.

* A NixOS test is also added that uses cups-pdf to generate pdf files, then checks if they contain the printed text.

  Note that currently, cups-pdf-to-pdf produces small and searchable pdf files, but the text is "garbled", i.e., characters are reassigned randomly.  This is a known issue and I don't know how to fix it:

  https://github.com/alexivkin/CUPS-PDF-to-PDF/issues/7

  So the test checks the pdf by rasterizing it, then applies OCR.

* As there is already a package request (see above),  this change might warrant a mention in the release notes.


###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


###### Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cups-pdf-to-pdf</li>
  </ul>
</details>


###### References

* [cups-pdf-to-pdf README](https://github.com/alexivkin/CUPS-PDF-to-PDF/blob/c14428c2ca8e95371daad7db6d11c84046b1a2d4/README.md)
* [cups-pdf-to-pdf in Repology](https://repology.org/project/cups-pdf-to-pdf/versions)
* [cups-pdf README](https://fossies.org/linux/cups-pdf/README)
* [cups-pdf in Repology](https://repology.org/project/cups-pdf/versions)

Yarny0 committed:
cups-pdf-to-pdf: init at unstable-2021-12-22

Note that cups-pdf refuses to run without root privileges.
To use the binary, one has to either convince cups to
call it with root privileges, or install it suid root.

Also note that currently, this cups-pdf-fork produces
small pdfs with selectable text, as promised.
However, copying the text produces "garbled" text
(characters are randomly reassigned).
This is a known issue and I don't know how to fix it:

https://github.com/alexivkin/CUPS-PDF-to-PDF/issues/7

Yarny0 committed:
nixos/cups-pdf: init

Some implementation notes:

* cups-pdf, and cups-pdf-to-pdf,
  support multiple instances with differing configurations.
  This can be accomplished by creating multiple configuration
  files with names `cups-pdf-{instance-name}.conf`.
  The Nixos module supports this feature by providing
  the option `instances` which is an attrset
  mapping instance names to instance configurations.
  To simplify module usage,
  an instance `pdf` is created by default.

* To use a cups-pdf instance, one also needs
  a cups queue that connects to the backend.
  The module does this automatically by default,
  using the `hardware.printers.ensurePrinters`.
  It uses one of the ppd files which is
  included in the cups-pdf package.
  If this isn't desired (e.g. because printer queues
  should be created by hand, or configured differently),
  the `installPrinter` option can be turned off
  (for each instance separately).

* In our configuration, cups calls external programs
  using the `cups` account and the `lp` group.
  cups-pdf refuses to operate without root privileges,
  likely because it needs to change the
  ownership of it output pdf files so that
  (only) the print job's owner can access them.

  The module installs a suid root wrapper for the backend
  program that can only be called by the `lp` group.
  The cups-pdf package is replaced by a wrapper
  package which calls the suid root wrapper.
  So cups can call its backend programs as usual.

Yarny0 committed:
cups-pdf: add vm test

Yarny0 committed:
cups-pdf: mention new package/module in 23.05 release notes

Yarny0 commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -54,6 +54,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [appvm](https://github.com/jollheef/appvm), Nix based app VMs. Available as [virtualisation.appvm](options.html#opt-virtualisation.appvm.enable).
 
+- [cups-pdf-to-pdf](https://github.com/alexivkin/CUPS-PDF-to-PDF), a pdf-generating cups backend based on [cups-pdf](https://www.cups-pdf.de/).  Available as [services.printing.cups-pdf](#opt-services.printing.cups-pdf.enable).
```
I don't think it makes sense to install `cups-pdf` and `cups-pdf-to-pdf` alongside on the same system: They are looking for their config files in the same place, so they couldn't be configured independently.  Hence I would refrain from ever adding two nixos options or modules for those packages.

(Of course we can patch the package to look for a different path, but I guess that would be a bit surprising for package users who do not use the NixOS module.)

If we want to provide the original `cups-pdf` package as an alternative, we can simply add a `package` option where the user can pick the favored package, something like

```nix
{
  type = lib.types.package;
  default = pkgs.cups-pdf-to-pdf;
  example = lib.literalExample "pkgs.cups-pdf";
}
```

to switch to the other package.  The NixOS module would then use this package to perform its suid wrapper magic.

However, since `cups-pdf` is a well-known name (at least, that's my feeling), and since `cups-pdf-to-pdf` is just a fork, I would still call this option `cups-pdf` and hope it leads to less confusion.

Yarny0 commented on nixos/doc/manual/from_md/release-notes/rl-2211.section.xml:
```diff
@@ -119,6 +119,15 @@
+          a pdf-generating cups backend based on
+          <link xlink:href="https://www.cups-pdf.de/">cups-pdf</link>.
+          Available as
+          <link linkend="opt-services.printing.cups-pdf.enable">services.printing.cups-pdf</link>.
```
(see above)

Yarny0 commented:
Dear @nrdsp ,

adding a `package` option should be easy.  Would you prefer we offer such an option?  Just tell me and I'll push an updated branch.  Do you think we should also package the original `cups-pdf` package?  I remember building this package also before I discovered the improved fork, and I faintly remember it was quite easy and no difference to the forked package.  So if you like `cups-pdf` packaged as well, just tell me and I'll add it to the branch, together with another vm test.

nrdsp reviewed (approved):
The package seems to build on my machine, even if it complained about some functions being deprecated ([opened issue upstream](https://github.com/alexivkin/CUPS-PDF-to-PDF/issues/9)).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1070

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -54,6 +54,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [appvm](https://github.com/jollheef/appvm), Nix based app VMs. Available as [virtualisation.appvm](options.html#opt-virtualisation.appvm.enable).
 
+- [cups-pdf-to-pdf](https://github.com/alexivkin/CUPS-PDF-to-PDF), a pdf-generating cups backend based on [cups-pdf](https://www.cups-pdf.de/).  Available as [services.printing.cups-pdf](#opt-services.printing.cups-pdf.enable).
```
```suggestion
- [cups-pdf-to-pdf](https://github.com/alexivkin/CUPS-PDF-to-PDF), a pdf-generating cups backend based on [cups-pdf](https://www.cups-pdf.de/). Available as [services.printing.cups-pdf](#opt-services.printing.cups-pdf.enable).
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+
+let
+
+  inherit (lib.attrsets) filterAttrs getAttr mapAttrs mapAttrsToList;
+  inherit (lib.meta) getExe;
+  inherit (lib.modules) mkAfter mkDefault mkIf;
+  inherit (lib.options) literalExpression mkEnableOption mkOption;
+  inherit (lib.strings) concatStrings;
+  inherit (lib.trivial) pipe;
+  inherit (lib.types) attrsOf int lines nullOr oneOf package path singleLineStr str submodule;
```
```suggestion
```
please use with lib or prefix the the functions with lib. Maintaining that list goes not have a real benefit.

Yarny0 replied:
I try to avoid `with` whereever possible (similar to wildcard imports in Python), so I added `lib.` prefixes.

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+  inherit (lib.trivial) pipe;
+  inherit (lib.types) attrsOf int lines nullOr oneOf package path singleLineStr str submodule;
+
+  # cups calls its backends as user `lp` (which is good!),
+  # but cups-pdf wants to be called as `root`,
+  # so it can change ownership of files.
+  # We add a suid wrapper and a wrapper script
+  # to trick cups into calling the suid wrapper.
```
```suggestion
  # cups calls its backends as user `lp` (which is good!),
  # but cups-pdf wants to be called as `root`, so it can change ownership of files.
  # We add a suid wrapper and a wrapper script to trick cups into calling the suid wrapper.
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+  # to trick cups into calling the suid wrapper.
+
+  # wrapper script that redirects calls to the suid wrapper
+  wrapper = pkgs.writeTextFile {
```
```suggestion
  cups-pdf-to-pdf-wrapper = pkgs.writeTextFile {
```

Yarny0 replied:
I'm just calling this one `cups-pdf-wrapper`, as I attempt to keep this module as general as possible (see also the discussion in https://github.com/NixOS/nixpkgs/pull/182360#discussion_r928051896 ).

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+  wrapped = pkgs.buildEnv {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapped";
+    # using wrapper as first path ensures the wrapper is used
+    paths = [ wrapper pkgs.cups-pdf-to-pdf ];
```
```suggestion
    paths = [ cups-pdf-to-pdf-wrapper pkgs.cups-pdf-to-pdf ];
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+  };
+
+  # wrapped cups-pdf package that uses the suid wrapper
+  wrapped = pkgs.buildEnv {
```
```suggestion
  cups-pdf-to-pdf-wrapped = pkgs.buildEnv {
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+      example = "root";
+      description = ''
+        User for anonymous PDF creation.
+        An empty value disables this feature.
```
```suggestion
        A null value disables this feature.
```
null is not equal to an empty string.

Yarny0 replied:
(Github is hiding this suggestion and most of the suggestions below when I'm not logged in, so I haven't seen this and haven't fixed them yet.  I'm preparing another update, but it might take some time.)

Yarny0 replied:
It is really an empty string that disables the feature (according to the cups-pdf.conf sample file).  If the option is omitted, then `nobody` is used by default (see below); if it is set but left empty, anonymous pdf creation is disabled.  

I reworded the description to make that clearer ("An empty string disables...").

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+      description = ''
+        User for anonymous PDF creation.
+        An empty value disables this feature.
+        Avoid the default value <literal>nobody</literal>
+        as it might cause security issues.
```
We should assert that.

Yarny0 replied:
I'm not sure what the best solution is here.  I now set the default to `root` which might annoy anonymous clients but not pose any security risk.  To avoid the `nobody` account being used, I removed the `nullOr` type.  This means one should not be able to assign print jobs to the `nobody` account unless one directly sets this option to `nobody`.

The description also no longer mentiones `nobody`.  Is it acceptable to allow the user to explicitely use the `nobody` group if s/he really wants to?

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+
+  instanceConfig = { name, config, ... }: {
+    options = {
+      enable = (mkEnableOption "this cups-pdf instance")
+               // { default = true; };
```
```suggestion
      enable = (mkEnableOption "this cups-pdf instance") // { default = true; };
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+               // { default = true; };
+      installPrinter = (mkEnableOption ''
+        a CUPS printer queue for this instance.
+        The queue will be named after the instance and will use
+        the <filename>CUPS-PDF_opt.ppd</filename> ppd file.
+        If this is disabled, you need to add
+        the queue yourself to use the instance
```
```suggestion
        The queue will be named after the instance and will use the <filename>CUPS-PDF_opt.ppd</filename> ppd file.
        If this is disabled, you need to add the queue yourself to use the instance
```
Such short lines are harder to read. We don't need to limit us to 60 characters. There is plenty of space left.

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+
+  config = mkIf cupsPdfCfg.enable {
+    services.printing.enable = true;
+    services.printing.drivers = [ wrapped ];
```
```suggestion
    services.printing.drivers = [ cups-pdf-to-pdf-wrapped ];
```

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,205 @@
+    # the cups module will install the default config file,
+    # be we don't need it and it would confuse cups-pdf
+    systemd.services.cups.preStart = mkAfter ''
+      rm --force /var/lib/cups/cups-pdf.conf
```
```suggestion
      rm -f /var/lib/cups/cups-pdf.conf
```

SuperSandro2000 commented on nixos/tests/cups-pdf.nix:
```diff
@@ -0,0 +1,39 @@
+        machine.wait_until_succeeds(f"su - alice -c 'pdfinfo /var/spool/cups-pdf-{name}/users/alice/*.pdf'")
+        machine.succeed(f"cp /var/spool/cups-pdf-{name}/users/alice/*.pdf /tmp/{name}.pdf")
+        machine.copy_from_vm(f"/tmp/{name}.pdf", "")
+        run(f"${pkgs.imagemagickBig}/bin/convert -density 300 $out/{name}.pdf $out/{name}.jpeg", shell=True, check=True)
```
Do we really need to use imagemagickBig or is imagemagick enough?

Yarny0 replied:
imagemagick (not big) is failing, apparently due to missing ghostscript support.  I added a comment about that.

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+stdenv.mkDerivation rec {
+  pname = "cups-pdf-to-pdf";
+  version = "unstable-2021-12-22";
+  src = fetchFromGitHub {
+    owner = "alexivkin";
+    repo = "CUPS-PDF-to-PDF";
+    rev = "c14428c2ca8e95371daad7db6d11c84046b1a2d4";
+    hash = "sha512-FP+GGq3hVAXrLzidAsp3Q1WNzZxrIDaQHN36+RTu1W09FlOfZnIVdw/uJHBMWSm03HIBPSwQmpmpW8KqRv6izQ==";
+  };
+  buildInputs = [ cups ];
```
```suggestion
  version = "unstable-2021-12-22";

  src = fetchFromGitHub {
    owner = "alexivkin";
    repo = "CUPS-PDF-to-PDF";
    rev = "c14428c2ca8e95371daad7db6d11c84046b1a2d4";
    hash = "sha512-FP+GGq3hVAXrLzidAsp3Q1WNzZxrIDaQHN36+RTu1W09FlOfZnIVdw/uJHBMWSm03HIBPSwQmpmpW8KqRv6izQ==";
  };

  buildInputs = [ cups ];
```

Please insert some new lines between options in the package.

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+    hash = "sha512-FP+GGq3hVAXrLzidAsp3Q1WNzZxrIDaQHN36+RTu1W09FlOfZnIVdw/uJHBMWSm03HIBPSwQmpmpW8KqRv6izQ==";
+  };
+  buildInputs = [ cups ];
+  # cups-pdf wants to run with root privileges;
+  # to facilitate the use of a suid root wrapper,
+  # we have to use the binary wrapper here
```
```suggestion
  # binary wrapper is required for the suid wrapper to function
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+    runHook preBuild
+    gcc -O9 -s cups-pdf.c -o cups-pdf -lcups
+    runHook postBuild
+  '';
+  installPhase = ''
+    runHook preInstall
+    install -D --target-directory=${placeholder "out"}/lib/cups/backend cups-pdf
+    install -D --target-directory=${placeholder "out"}/etc/cups --mode=0644 cups-pdf.conf
+    install -D --target-directory=${placeholder "out"}/share/cups/model --mode=0644 *.ppd
+    runHook postInstall
+  '';
+  # without coreutils, cups-pdf would complain about missing `cp`
```
```suggestion
  '';

  installPhase = ''
    runHook preInstall
    install -D -t $out/lib/cups/backend cups-pdf
    install -Dm644 -t $out/etc/cups cups-pdf.conf
    install -Dm644 -t $out/share/cups/model *.ppd
    runHook postInstall
  '';

  # without coreutils, cups-pdf would complain about missing `cp`
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+    runHook postInstall
+  '';
+  # without coreutils, cups-pdf would complain about missing `cp`
+  preFixup = ''
+    wrapProgram ${placeholder "out"}/lib/cups/backend/cups-pdf  \
+        --prefix PATH : "${lib.strings.makeBinPath [ coreutils ]}"
+  '';
+  passthru.tests.vmtest = nixosTests.cups-pdf;
```
```suggestion

  passthru.tests.vmtest = nixosTests.cups-pdf;

```

instead of wrapping we can just replace https://github.com/alexivkin/CUPS-PDF-to-PDF/blob/master/cups-pdf.c#L1029 with ``${coreutils}/bin/cp``.

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+  passthru.tests.vmtest = nixosTests.cups-pdf;
+  meta = rec {
+    description = "A CUPS backend that turns print jobs into searchable PDF files";
+    downloadPage = homepage;
```
```suggestion
```
if homepage and downloadPage are equal we do not need to set this.

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+        --prefix PATH : "${lib.strings.makeBinPath [ coreutils ]}"
+  '';
+  passthru.tests.vmtest = nixosTests.cups-pdf;
+  meta = rec {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+    description = "A CUPS backend that turns print jobs into searchable PDF files";
+    downloadPage = homepage;
+    homepage = "https://github.com/alexivkin/CUPS-PDF-to-PDF";
+    license = lib.licenses.gpl2Only;
+    maintainers = [ lib.maintainers.yarny ];
```
```suggestion
    license = licenses.gpl2Only;
    maintainers = [ maintainers.yarny ];
```

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+      https://www.cups-pdf.de/
+
+      ${pname} is a fork of cups-pdf which tries hard to preserve
```
```suggestion
      cups-pdf-to-pdf is a fork of cups-pdf which tries hard to preserve
```
substitutung pname wherever possible is not improving code quality

Yarny0 commented:
Thanks @SuperSandro2000 for your suggestions.  The most recent branch should now incorporate all of them.

Yarny0 commented:
> [SuperSandro2000](https://github.com/SuperSandro2000) added the [2.status: merge conflict](https://github.com/NixOS/nixpkgs/labels/2.status%3A%20merge%20conflict) label [3 days ago](https://github.com/NixOS/nixpkgs/pull/182360#event-7189852229)

Rebased onto current `nixos-unstable` ( 762b003329510ea855b4097a37511eb19c7077f0 ) and resolved the conflict (just had to rerun `md-to-db.sh`).  cups-pdf vm test still passes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/52

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,60 @@
+    owner = "alexivkin";
+    repo = "CUPS-PDF-to-PDF";
+    rev = "c14428c2ca8e95371daad7db6d11c84046b1a2d4";
+    hash = "sha512-FP+GGq3hVAXrLzidAsp3Q1WNzZxrIDaQHN36+RTu1W09FlOfZnIVdw/uJHBMWSm03HIBPSwQmpmpW8KqRv6izQ==";
```
Please use the standard sha256

SuperSandro2000 commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,60 @@
+  buildInputs = [ cups ];
+
+  postPatch = ''
+    sed -r 's|(gscall, size, ")cp |\1${lib.strings.makeBinPath [ coreutils ]}/cp |' cups-pdf.c -i
```
```suggestion
    sed -r 's|(gscall, size, ")cp |\1${coreutils}/bin/cp |' cups-pdf.c -i
```

SuperSandro2000 commented on nixos/tests/cups-pdf.nix:
```diff
@@ -0,0 +1,40 @@
+        machine.wait_until_succeeds(f"su - alice -c 'pdfinfo /var/spool/cups-pdf-{name}/users/alice/*.pdf'")
+        machine.succeed(f"cp /var/spool/cups-pdf-{name}/users/alice/*.pdf /tmp/{name}.pdf")
+        machine.copy_from_vm(f"/tmp/{name}.pdf", "")
+        run(f"${lib.makeBinPath [ pkgs.imagemagickBig ]}/convert -density 300 $out/{name}.pdf $out/{name}.jpeg", shell=True, check=True)
```
```suggestion
        run(f"${pkgs.imagemagickBig}/bin/convert -density 300 $out/{name}.pdf $out/{name}.jpeg", shell=True, check=True)
```
Do we really need imagemagickBig? Or could we use a smaller variant?

SuperSandro2000 commented on nixos/tests/cups-pdf.nix:
```diff
@@ -0,0 +1,40 @@
+        machine.succeed(f"cp /var/spool/cups-pdf-{name}/users/alice/*.pdf /tmp/{name}.pdf")
+        machine.copy_from_vm(f"/tmp/{name}.pdf", "")
+        run(f"${lib.makeBinPath [ pkgs.imagemagickBig ]}/convert -density 300 $out/{name}.pdf $out/{name}.jpeg", shell=True, check=True)
+        assert text.encode() in run(f"${lib.makeBinPath [ pkgs.tesseract ]}/tesseract $out/{name}.jpeg stdout", shell=True, check=True, capture_output=True).stdout
```
```suggestion
        assert text.encode() in run(f"${pkgs.tesseract}/bin/tesseract $out/{name}.jpeg stdout", shell=True, check=True, capture_output=True).stdout
```

SuperSandro2000 reviewed (commented):
makeBinPath should only be used with multiple packages when you need a PATH style variable.

Yarny0 commented:
Thanks for your suggestions, @SuperSandro2000 .  I applied all of them, up to changing the Imagemagick package:
We need Ghostscript support in order to rasterize the pdf file, and the "Big" variants are the only ones satisfying that (according to `all-packages.nix`).

I also rebased onto current `nixos-unstable` and resolved the conflict.  Finally, I converted the NixOS module to markdown documentation.  VM test still passes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/printing-to-pdf-within-wine/23359/2

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+  # but cups-pdf wants to be called as `root`, so it can change ownership of files.
+  # We add a suid wrapper and a wrapper script to trick cups into calling the suid wrapper.
+
+  # wrapper script that redirects calls to the suid wrapper
+  cups-pdf-wrapper = pkgs.writeTextFile {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapper.sh";
+    executable = true;
+    destination = "/lib/cups/backend/cups-pdf";
+    checkPhase = ''
+      ${pkgs.stdenv.shellDryRun} "$target"
+      ${lib.getExe pkgs.shellcheck} "$target"
+    '';
+    text = ''
+      #! ${pkgs.runtimeShell}
+      exec "${config.security.wrapperDir}/cups-pdf" "$@"
+    '';
+  };
```
Why do we need this wrapper script if we already have a wrapper? Why can't we use the suid wrapper alone?

Yarny0 replied:
CUPS is looking for its backend binaries in `$CUPS_PATH/lib/cups/backend/`, but the suid wrapper is placed in `/run/wrappers/bin/cups-pdf`.  So something must redirect CUPS from the former to the latter place.  I also tried this with a symlink, i.e.

```
$CUPS_PATH/lib/cups/backend/cups-pdf -> /run/wrappers/bin/cups-pdf
```

but CUPS complains about this having insecure permissions and refuses to execute it.  Putting a suid wrapper directly into `$CUPS_PATH/lib/cups/backend/` is also not possible since that is a store path.  So apart from completely reorganizing the way CUPS finds its backend binaries, using two wrappers is the only method I can come up with.

I added a note about that implementation detail in the commit message of the corresponding commit so whoever wants to find out how I came up with this construction can look it up.

SuperSandro2000 replied:
> I added a note about that implementation detail in the commit message of the corresponding commit so whoever wants to find out how I came up with this construction can look it up.

Please move/copy this into a code, otherwise this question will come up again.

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+    '';
+  };
+
+  # wrapped cups-pdf package that uses the suid wrapper
+  cups-pdf-wrapped = pkgs.buildEnv {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapped";
+    # using the wrapper as first path ensures it is used
+    paths = [ cups-pdf-wrapper pkgs.cups-pdf-to-pdf ];
+    ignoreCollisions = true;
+  };
```
same here

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+
+  instanceSettings = name: {
+    freeformType = lib.types.nullOr
+      (lib.types.oneOf [ lib.types.int lib.types.str lib.types.path lib.types.package ]);
```
```suggestion
      (lib.types.oneOf with lib.types; [ int str path package ]);
```

Yarny0 replied:
(see below)

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+    # override defaults:
+    # inject instance name into paths,
+    # also avoid conflicts between user names and special dirs
+    options.Out = lib.mkOption {
+      type = lib.types.nullOr lib.types.singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/users/\${USER}";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/users/\${USER}";
+      example = "\${HOME}/cups-pdf";
+      description = lib.mdDoc "output directory";
+    };
+    options.AnonDirName = lib.mkOption {
+      type = lib.types.nullOr lib.types.singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/anonymous";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp";
+      description = lib.mdDoc "path for anonymously created PDF files";
+    };
+    options.Spool = lib.mkOption {
+      type = lib.types.nullOr lib.types.singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/spool";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp/cups-pdf";
+      description = lib.mdDoc "spool directory";
+    };
+    options.Anonuser = lib.mkOption {
+      type = lib.types.singleLineStr;
+      default = "root";
+      description = lib.mdDoc ''
+        User for anonymous PDF creation.
+        An empty string disables this feature.
+      '';
+    };
```
We should use some with lib.types to reduce the duplication here.

Yarny0 replied:
done

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+      type = lib.types.nullOr lib.types.singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/anonymous";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp";
```
/tmp is a bad example. We should suggest something under /var/lib.

Yarny0 replied:
done

SuperSandro2000 commented on nixos/modules/services/printing/cups-pdf.nix:
```diff
@@ -0,0 +1,180 @@
+      type = lib.types.nullOr lib.types.singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/users/\${USER}";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/users/\${USER}";
+      example = "\${HOME}/cups-pdf";
```
Isn't this relative to root or your current user?

Yarny0 commented:
Thanks again for your review, @SuperSandro2000 .  I hope I could answer all your questions.  The brach is also rebased onto current `nixos-unstable`.  The vm test still passes.

RaitoBezarius commented:
Can you move the release notes to 23.05 ?

Yarny0 commented:
> Can you move the release notes to 23.05 ?

Done.  Also rebased onto current `nixos-unstable`.  Test still passes.

RaitoBezarius commented on pkgs/misc/cups/drivers/cups-pdf-to-pdf/default.nix:
```diff
@@ -0,0 +1,60 @@
+  # https://fossies.org/linux/cups-pdf/README
+  buildPhase = ''
+    runHook preBuild
+    gcc -O9 -s cups-pdf.c -o cups-pdf -lcups
```
This fails on Darwin, which is definitely something that can be used there I think.

Yarny0 replied:
Hmmmm, I don't know much about Darwin.  Following the first hint in https://nixos.org/manual/nixpkgs/stable/#sec-darwin , I replaced `gcc` with `$CC`.  Hopefully, this calls `clang` and the command line still works.  On `x86_64-linux`, the package still builds fine with this change (and the test passes).

Can you try if this works on Darwin.

RaitoBezarius replied:
According to OfBorg, this works on Darwin: https://logs.nix.ci/?key=nixos/nixpkgs.182360&attempt_id=e17af287-4a25-4e81-bc11-09cf5d8e3f13

RaitoBezarius commented:
Alas, release notes conflicts :'( ; please fix the conflicts and I think we are good to merge.

Yarny0 commented:
Rebased onto current `nixos-unstable`, merge conflict is resolved.  Test still passes.

Yarny0 commented:
Rebased onto `nixos-unstable` again, due to a new merge conflict.  Test still passes.

RaitoBezarius commented:
BTW, naive question but, you wrote:

> Also note that currently, this cups-pdf-fork produces
> small pdfs with selectable text, as promised.
> However, copying the text produces "garbled" text
> (characters are randomly reassigned).
> This is a known issue and I don't know how to fix it:

Does that mean this package provide zero value with respect to its main feature?

Yarny0 commented:
> BTW, naive question but, you wrote:
> > [...]
> 
> Does that mean this package provide zero value with respect to its main feature?

This depends on what is considered the main feature, but if the main feature is text-searchable pdf-files, then sadly *yes*.
The original [cups-pdf](https://www.cups-pdf.de) produces rasterized pdfs, thus the resolution is limited by the rasterization when viewed/printed on high-dpi devices.  Even without the possibility to extract text, the pdfs from [CUPS-PDF-to-PDF](https://github.com/alexivkin/CUPS-PDF-to-PDF) are usually vector graphics and ofter require less space.  Depending on the use-case, this certainly can be an improvement (it is for me).

RaitoBezarius commented:
This is a good candidate for a more general nixpkgs warning (https://github.com/NixOS/rfcs/pull/127), but I think it's fine.
